### PR TITLE
CI: refactor entrypoints location

### PR DESCRIPTION
**Pull request by kpouget**
This commit refactors the location of the CI entrypoints.

One goal is to be able to trigger CI tests locally, without having
issues with `PATH` references (files used to be stored in
`/usr/local/bin`.

The other goal is to be able to trigger custom tests in the CI, and
store permanently the test routes. Such tests will be specified with
the following anchor in the last commit of the PR:

test-path: echo hello world

Note that `test-path` must point to a script in `testing/` (to avoid
running arbitrary commands not shown in the PR file diff).

The execution of *all* the test paths will be triggered with `\test
custom` in the PR message.

### CODE REVIEW

**Comment by kpouget** [comment]
Hello @omertuc, 

can you please have a look at this PR, let me know what you think about it?
the [revert](https://github.com/openshift-psap/ci-artifacts/blob/master/.github/workflows/revert-testing.yml) idea cannot work because the Action cannot push into `master`, so I'll add a commit to delete it,

and I like better this new design, the test script remains in the repo, and the testing is triggered via a `test-path:` mention in the [top-most] commit, + `/test custom` triggered in the PR. See [this](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/255/pull-ci-openshift-psap-ci-artifacts-master-custom/1437671080700416000/artifacts/custom/custom-hook/build-log.txt) `hello world` for an illustration.

@pacevedom FYI, feel free to comment if you have any concern or suggestion!

I'll add a bit of code in this PR to
1. delete the `revert` Github Action
2. store the logs of each `test-path` into a dedicated `artifacts` file (`stderr` and `stdout`)
3. store OCP and `ci-artifacts` version in the `artifacts`, so that all operators can benefit from this `ci-dashboard` [PR](https://github.com/openshift-psap/ci-dashboard/pull/34)

**Comment by kpouget** [comment]
@dagrayvid @omertuc @ArangoGutierrez any feedback on this PR?

File: build/Dockerfile [hunk 1]
```diff
@@ -61,6 +58,12 @@ ENV USER_NAME=psap-ci-runner \
 
 COPY . ${HOME}/
 
+# Prepare the CI `run` entrypoint
+RUN echo -e '#!/usr/bin/env bash \n\
```
Comment (omertuc) [inline build/Dockerfile:62]: I guess it's too early to use https://www.docker.com/blog/introduction-to-heredocs-in-dockerfiles/

File: testing/custom/test.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#! /usr/bin/env bash
+
+exec echo "Hello World"
```
Comment (omertuc) [inline testing/custom/test.sh:3]: Is this file commited on purpose?

File: testing/nightly/cluster.sh [hunk 3]
```diff
@@ -27,12 +30,8 @@ case ${action:-} in
         cluster_upgrade
         exit 0
         ;;
-    -*)
-        echo "FATAL: Unknown option: ${action}"
-        exit 1
-        ;;
     *)
-        echo "FATAL: Nothing to do ..."
+        echo "FATAL: Nothing to do ... (action='$action')"
```
Comment (omertuc) [inline testing/nightly/cluster.sh:34]: Maybe change message to "Unsupported $action"

File: testing/nightly/custom.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
+#! /usr/bin/env bash
+
```
Comment (omertuc) [inline testing/nightly/custom.sh:2]: Could you add a description at the top of the file explaining what it's doing (overview)?

File: testing/custom/test.sh (hunk 2 above)
Comment (kpouget) [reply testing/custom/test.sh:3]: it's here for `/test custom` to work, same as this commit https://github.com/openshift-psap/ci-artifacts/pull/255/commits/6311ca474d12282d4e7b2054190812afa723862d :#

not sure if I should remove it or not :/

File: build/Dockerfile (hunk 1 above)
Comment (kpouget) [reply build/Dockerfile:62]: I tried the multiline ```
echo ' ...
...
...'
```
but it failed, not sure if this one has a chance to be supported ... ?

File: testing/nightly/cluster.sh (hunk 3 above)
Comment (kpouget) [reply testing/nightly/cluster.sh:34]: fixed

File: testing/nightly/custom.sh (hunk 4 above)
Comment (kpouget) [reply testing/nightly/custom.sh:2]: fixed:
```
# This script is executed from `ci-artifacts` PR, when requesting
# `\test custom`.
#
# It looks at the message of the last commit of the PR (2nd parent of
# HEAD merge commit), and scans it for a test-path directive:
# > test-path: testing/.../test.sh [flags]
# and executes the test command.
#
# Calling this script on a commit without a `test-path` fails the test
# (exit 1).
```

File: testing/nightly/custom.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+set -o pipefail
+set -o errexit
+set -o nounset
+
+# This script is executed from `ci-artifacts` PR, when requesting
+# `\test custom`.
+#
+# It looks at the message of the last commit of the PR (2nd parent of
+# HEAD merge commit), and scans it for a test-path directive:
```
Comment (omertuc) [inline testing/nightly/custom.sh:11]: s/scans it/scans its commit message/

?

File: testing/nightly/custom.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+BASE_ARTIFACT_DIR=${ARTIFACT_DIR}
+
+while read cmd;
+do
+    if ! grep "^testing/" -q <<< "$cmd"; then
+        echo "ERROR: test command must start with testing/: $cmd"
+        continue
+    fi
+
+    if grep ";" -q <<< "$cmd"; then
```
Comment (omertuc) [inline testing/nightly/custom.sh:54]: This is definitely insufficient from a security perspective, there are a lot of ways to bypass this. This is a battle not worth fighting - we must make sure only authorized users can have CI run on their PRs

Comment (kpouget) [reply testing/nightly/custom.sh:54]: this block is not necessary anymore, I removed the `eval` function that used to be called, and use instead:
```
bash $(echo "$cmd") 
```
my goal was that arbitrary code could not be executed from the commit message, eg
```
test-path: rm -rf /
```
fails with
```
/usr/bin/rm: /usr/bin/rm: cannot execute binary file
```
so only files part of the repo can get executed, and hence already in the repo or part of the PR diff

I'll remove this block

**Comment by kpouget** [comment]
> /lgtm

thanks Omer, I'll trigger a few tests before merging this one

/test ?

**Comment by kpouget** [comment]
`sro-e2e-master` failed because of [this](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/255/pull-ci-openshift-psap-ci-artifacts-master-sro-e2e-master/1438830052841099264/artifacts/e2e-master/nightly/artifacts/007__sro__run_e2e_test/sro_e2e_test.log), which isn't related to this PR
```
2021-09-17T12:03:05.384Z	INFO	[1;34mdeploy  [0m	Created	{"Kind": "SpecialResource", "Name": "simple-kmod"}
STEP: Waiting for simple-kmod daemonset to be ready
STEP: Waiting for simple-kmod daemonset to be ready
panic: test timed out after 40m0s
```

/approve

**Pull request merged**